WebXR Test API Hit Test extensions

fujii opened the pull request:
#### 0d8281c39722575779f9b4fcdd97e34afd3df055
<pre>
WebXR Test API Hit Test extensions
<a href="https://bugs.webkit.org/show_bug.cgi?id=301463">https://bugs.webkit.org/show_bug.cgi?id=301463</a>

Reviewed by Sergio Villar Senin.

Implemented the Hit Test extensions of WebXR Test API by porting
Chromium webxr-test.js to C++.

Found out and fixed a bug. As per the spec, in requestHitTestSource
and requestHitTestSourceForTransientInput, if the entityTypes was not
provided, the effective entityTypes should be set to an array
containing single element, &quot;plane&quot;.

Tests: imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_transientInputSources.https.html

* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_inputSources.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_refSpaces.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_transientInputSources.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_unlocalizable.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/xrViewport_valid.https-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webxr/webxr_availability.http.sub-expected.txt: Removed a duplicated expectation.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::entityTypesFromOptions):
(WebCore::WebXRSession::requestHitTestSource):
(WebCore::WebXRSession::requestHitTestSourceForTransientInput):
* Source/WebCore/dom/DOMPointInit.idl:
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
* Source/WebCore/testing/FakeXRWorldInit.h: Added.
* Source/WebCore/testing/FakeXRWorldInit.idl: Added.
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):
(WebCore::SimulatedXRDevice::requestHitTestSource):
(WebCore::SimulatedXRDevice::requestTransientInputHitTestSource):
(WebCore::SimulatedXRDevice::hitTestWorld):
(WebCore::SimulatedXRDevice::setWorld):
(WebCore::SimulatedXRDevice::clearWorld):
(WebCore::WebFakeXRDevice::setWorld):
(WebCore::WebFakeXRDevice::clearWorld):
* Source/WebCore/testing/WebFakeXRDevice.h:
* Source/WebCore/testing/WebFakeXRDevice.idl:
* Source/WebCore/testing/WebXRTest.cpp:
(WebCore::WebXRTest::simulateDeviceConnection):
* Source/WebCore/testing/WebXRTest.h:
* Source/WebCore/testing/WebXRTest.idl:

Canonical link: <a href="https://commits.webkit.org/303898@main">https://commits.webkit.org/303898@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b75dd1f62ccc5898c4a94ddfcf43a52493e5e88

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133708 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6218 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44888 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141281 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85767 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135578 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6740 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6081 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102277 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69630 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136655 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4758 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119861 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83079 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4637 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2252 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113777 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37978 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143929 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5887 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38560 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110659 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5971 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5029 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110847 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4499 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116117 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59634 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20696 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5940 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34422 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5786 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69404 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6031 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5894 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

svillar commented on Source/WebCore/Modules/webxr/WebXRFrame.cpp:
```diff
@@ -418,14 +415,12 @@ ExceptionOr<Vector<Ref<WebXRTransientInputHitTestResult>>> WebXRFrame::getHitTes
-        Vector<Ref<WebXRHitTestResult>> hitTestResults(platformResult.results.size());
-        for (auto platformHitTestResult : platformResult.results)
-            hitTestResults.append(WebXRHitTestResult::create(*this, platformHitTestResult));
+        Vector<Ref<WebXRHitTestResult>> hitTestResults = platformResult.results.map([&](auto& platformHitTestResult) { return WebXRHitTestResult::create(*this, platformHitTestResult); });
```
What are the changes in this file for? Unless I misunderstand them, they're basically doing the same than the current code...

fujii replied:
The previous code was crashing. For example,

~~~
Vector<T> results(2);
results.append(x);
~~~

This code adds two empty items and x to results.

svillar replied:
You can always reserve capacity with Vector<T, capacity> results. Not sure if this should be part of this patch BTW, seems a bug fix that perhaps deserve a different PR?

fujii replied:
There are two capacities, inline capacity and initial capacity. You are talking about inline capacity. It's has to be a constant.
OK. I'm going to split the bug fix parts.

svillar commented on Source/WebCore/Modules/webxr/WebXRSession.cpp:
```diff
@@ -760,6 +760,14 @@ bool WebXRSession::isHandTrackingEnabled() const
+{
+    if (options.entityTypes.isEmpty())
+        return { XRHitTestTrackableType::Plane };
+    return options.entityTypes;
```
What about using ternary operator here? `return options.entityTypes.isEmpty() ? { ...} : options.entityTypes`

fujii replied:
I need `Vector<XRHitTestTrackableType>` before the `{ ... }`.
~~~
 return options.entityTypes.isEmpty() ? Vector<XRHitTestTrackableType> { XRHitTestTrackableType::Plane } : options.entityTypes;
~~~
This doesn't look nice to me.

svillar commented on Source/WebCore/testing/WebFakeXRDevice.cpp:
```diff
@@ -233,6 +264,241 @@ void SimulatedXRDevice::deleteTransientInputHitTestSource(PlatformXR::TransientI
+                }
+                return { position, orientation };
+            };
+            constexpr double epsilon = 0.001;
```
I am not sure we need all this trigonometry stuff. After checking TransformationMatrix.cpp I see there dot/cross products and matrix decompositions, are we sure we need to reimplement them here?

fujii replied:
FloatPoint3D has only 3 elements, x, y and z. But, these code uses 4 elements x, y, z and w. That makes it unique.
However, decomposition of TransformationMatrix class can be used. Fixed.

svillar replied:
That's because it's a quaternion, but there is a Quaternion.h in platform

fujii replied:
In my understanding, it's Homogeneous Coordinates.
https://en.wikipedia.org/wiki/Homogeneous_coordinates
https://en.wikipedia.org/wiki/Quaternion

svillar replied:
> In my understanding, it's Homogeneous Coordinates. https://en.wikipedia.org/wiki/Homogeneous_coordinates https://en.wikipedia.org/wiki/Quaternion

Yeah the ones in the matrix are homogeneous coordinates, I was talking about FloatPoint3D vs Quaternion. Sorry for the confusion

svillar reviewed (approved):
r+

fujii committed:
WebXR Test API Hit Test extensions
https://bugs.webkit.org/show_bug.cgi?id=301463

Reviewed by Sergio Villar Senin.

Implemented the Hit Test extensions of WebXR Test API by porting
Chromium webxr-test.js to C++.

Found out and fixed a bug. As per the spec, in requestHitTestSource
and requestHitTestSourceForTransientInput, if the entityTypes was not
provided, the effective entityTypes should be set to an array
containing single element, "plane".

Tests: imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_transientInputSources.https.html

* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_inputSources.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_refSpaces.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_transientInputSources.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/hit-test/ar_hittest_subscription_unlocalizable.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webxr/xrViewport_valid.https-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/webxr/webxr_availability.http.sub-expected.txt: Removed a duplicated expectation.
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::entityTypesFromOptions):
(WebCore::WebXRSession::requestHitTestSource):
(WebCore::WebXRSession::requestHitTestSourceForTransientInput):
* Source/WebCore/dom/DOMPointInit.idl:
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:
* Source/WebCore/testing/FakeXRWorldInit.h: Added.
* Source/WebCore/testing/FakeXRWorldInit.idl: Added.
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):
(WebCore::SimulatedXRDevice::requestHitTestSource):
(WebCore::SimulatedXRDevice::requestTransientInputHitTestSource):
(WebCore::SimulatedXRDevice::hitTestWorld):
(WebCore::SimulatedXRDevice::setWorld):
(WebCore::SimulatedXRDevice::clearWorld):
(WebCore::WebFakeXRDevice::setWorld):
(WebCore::WebFakeXRDevice::clearWorld):
* Source/WebCore/testing/WebFakeXRDevice.h:
* Source/WebCore/testing/WebFakeXRDevice.idl:
* Source/WebCore/testing/WebXRTest.cpp:
(WebCore::WebXRTest::simulateDeviceConnection):
* Source/WebCore/testing/WebXRTest.h:
* Source/WebCore/testing/WebXRTest.idl:

Canonical link: https://commits.webkit.org/303898@main
